### PR TITLE
[ISSUE-304][BUG]HA port being occupied makes master cannot normally launch

### PR DESCRIPTION
**Pull request by nafiyAix**
### What changes were proposed in this pull request?
Ratis HA port being occupied makes master cannot normally launch

### Why are the changes needed?
In original logic, it doesn't check if the port is occupied here, it will only check if the port is located in the valid port range.
image
And if the HA port is occupied (images pasted in Issue), it will throw below exception during Ratis server’s learder election, then make the Master's process stuck there and block the other component's initialization such as prometheus http endpoint.
So it is necessary to check whether the HA is occupied before the Ratis server launch, and exit the Master process if it is.

After changed, Master will throw below exception if HA port is occupied.
```
starting com.aliyun.emr.rss.service.deploy.master.Master, logging to /var/log/rss/rss-root-com.aliyun.emr.rss.service.deploy.master.xxxx.xxxr.xxx.io.out
failed to launch: nice -n 0 /usr/share/rss/bin/rss-class com.aliyun.emr.rss.service.deploy.master.Master
  Exception in thread "main" java.lang.IllegalArgumentException: This machine's Ratis port 9872 has been occupied, please refer to configuration guide to set it as another port in config file for all nodes.
  	at com.aliyun.emr.rss.service.deploy.master.clustermeta.ha.HANodeDetails.throwConfException(HANodeDetails.java:151)
  	at com.aliyun.emr.rss.service.deploy.master.clustermeta.ha.HANodeDetails.loadHAConfig(HANodeDetails.java:110)
  	at com.aliyun.emr.rss.service.deploy.master.clustermeta.ha.MetaHandler.setUpMasterRatisServer(MetaHandler.java:44)
  	at com.aliyun.emr.rss.service.deploy.master.Master.<init>(Master.scala:56)
  	at com.aliyun.emr.rss.service.deploy.master.Master$.main(Master.scala:555)
  	at com.aliyun.emr.rss.service.deploy.master.Master.main(Master.scala)
full log in /var/log/rss/rss-root-com.aliyun.emr.rss.service.deploy.master.xxx.xxx.xxx.io.out
```
### Related issues.
#304 

### How was this patch tested?

/cc @AngersZhuuuu 

/assign @main-reviewer

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
pls update the exception stack in pr desc after this path

**Comment by nafiyAix** [comment]
> 

updated

**Comment by nafiyAix** [comment]
> pls update the exception stack in pr desc after this path

updated

**Comment by AngersZhuuuu** [comment]
ping @FMX @waitinfuture

File: server-master/src/main/java/com/aliyun/emr/rss/service/deploy/master/clustermeta/ha/HANodeDetails.java [hunk 1]
```diff
@@ -160,4 +169,27 @@ public static boolean isLocalAddress(InetAddress addr) {
 
     return local;
   }
+
+  public static boolean isLocalPortOccupied(int port) {
+    ServerSocket ss = null;
+    DatagramSocket ds = null;
+    try {
+      ss = new ServerSocket(port);
```
Comment (FMX) [inline server-master/src/main/java/com/aliyun/emr/rss/service/deploy/master/clustermeta/ha/HANodeDetails.java:177]: I think it is dangerous to occupy a port to check its availability. And you'll never know what might happen after you release a port.

**Comment by FMX** [comment]
I think we might need to catch an exception at the ratis server starting function and terminate the master process if the ratis port is unavailable.

``` 
Meta.scala
 metaSystem.getRatisServer().start();
```

File: server-master/src/main/java/com/aliyun/emr/rss/service/deploy/master/clustermeta/ha/HANodeDetails.java (hunk 1 above)
Comment (nafiyAix) [reply server-master/src/main/java/com/aliyun/emr/rss/service/deploy/master/clustermeta/ha/HANodeDetails.java:177]: Agree, and may i know is there any other way to check if the port is usable without actually occupying it?

**Comment by nafiyAix** [comment]
> I think we might need to catch an exception at the ratis server starting function and terminate the master process if the ratis port is unavailable.
> 
> ```
> Meta.scala
>  metaSystem.getRatisServer().start();
> ```
There are the exceptions would be caught when using occupied HA port to launch ratis, but seems both of them has been caught inside the Ratis. 
FYI: The 1st exception will thrown multiple times than the 2nd exception will be thrown.
1. WARN [nioEventLoopGroup-4-1] DefaultChannelPipeline: An exceptionCaught() event was fired, and it reached at the tail
of the pipeline. It usually means the last handler in the pipeline did not handle the exception.
java.lang.IllegalArgumentException: Reply case not set in proto: RAFTNETTYSERVERREPLY_NOT_SET
        at org.apache.ratis.netty.NettyRpcProxy.getCallId(NettyRpcProxy.java:94)
        at org.apache.ratis.netty.NettyRpcProxy$Connection$1.channelRead0(NettyRpcProxy.java:115)
        at org.apache.ratis.netty.NettyRpcProxy$Connection$1.channelRead0(NettyRpcProxy.java:109)
2. ERROR [master2@group-F9C189140641-LeaderElection1] LeaderElection: master2@group-F9C189140641-LeaderElection1: Failed,
 state=RUNNING
java.lang.IllegalStateException: id is empty.
        at org.apache.ratis.util.Preconditions.assertTrue(Preconditions.java:45)
        at org.apache.ratis.protocol.RaftPeerId.<init>(RaftPeerId.java:67)
        at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1660)
        at org.apache.ratis.protocol.RaftPeerId.valueOf(RaftPeerId.java:41)
        at org.apache.ratis.server.impl.LeaderElection.waitForResults(LeaderElection.java:391)
        at org.apache.ratis.server.impl.LeaderElection.submitRequestAndWaitResult(LeaderElection.java:283)
        at org.apache.ratis.server.impl.LeaderElection.askForVotes(LeaderElection.java:307)
        at org.apache.ratis.server.impl.LeaderElection.run(LeaderElection.java:236)
        at java.lang.Thread.run(Thread.java:748)

I tried to launch a thread to monitor the server state by below methods and shut down master once the Ratis Server’s state becomes CLOSING or CLOSED after the 2nd exception below thrown, but it seems the 2nd exception would only change the state of LeaderElection and the Server's State would still be NEW which won't trigger the Master process shutdown. May I know is there a way to get the LeaderElection externally or other suggestions?
cc: @FMX @AngersZhuuuu 
```
metaSystem.getRatisServer().getServerState()
```

File: server-master/src/main/java/com/aliyun/emr/rss/service/deploy/master/clustermeta/ha/HANodeDetails.java (hunk 1 above)
Comment (waitinfuture) [reply server-master/src/main/java/com/aliyun/emr/rss/service/deploy/master/clustermeta/ha/HANodeDetails.java:177]: Even if we found a way to check port availability here, we can't guarantee that the port will not be occupied before ratis uses it. I think we should check what happens when port is unavailable but ratis server starts, better to throw Exception and exit then.

**Comment by FMX** [comment]
@nafiyAix  I think a raft server is essential for an RSS master to work as desired. If a raft server fails to start, the worker should not start either. So my solution is quite straightforward, I think changing a method of MetaHandler should be enough. 

Here is an example.

```java
public void setUpMasterRatisServer(RssConf conf) throws IOException {
    HANodeDetails haNodeDetails = HANodeDetails.loadHAConfig(conf);
    metaSystem.setRatisServer(
        HARaftServer.newMasterRatisServer(
            this, conf,
            haNodeDetails.getLocalNodeDetails(),
            haNodeDetails.getPeerNodeDetails()));
    try {
      metaSystem.getRatisServer().start();
    } catch (IOException e) {
      if(e.getCause() instanceof BindException) {
        LOG.error("Raft port is used, master will abort", e);
        System.exit(-1);
      }else{
        LOG.error("Unknown exception occurred in starting raft server",e);
      }
    }
  }
```

I tried on a host and I get the result.

> 22/08/15 15:20:15,867 INFO [node1-impl-thread1] JmxRegister: Successfully registered JMX Bean with object name Ratis:service=RaftServer,group=group-085C98AE2908,id=node1
> 22/08/15 15:20:15,870 INFO [main] RaftServer: node1: start RPC server
> 22/08/15 15:20:15,871 ERROR [main] MetaHandler: Raft port is used, master will abort
> java.io.IOException: node1: Failed to start NettyRpcService
> 	at org.apache.ratis.netty.server.NettyRpcService.startImpl(NettyRpcService.java:141)
> 	at org.apache.ratis.util.LifeCycle.startAndTransition(LifeCycle.java:270)
> 	at org.apache.ratis.server.RaftServerRpcWithProxy.start(RaftServerRpcWithProxy.java:72)
> 	at org.apache.ratis.server.impl.RaftServerProxy.lambda$start$6(RaftServerProxy.java:383)
> 	at org.apache.ratis.util.LifeCycle.startAndTransition(LifeCycle.java:270)
> 	at org.apache.ratis.server.impl.RaftServerProxy.start(RaftServerProxy.java:381)
> 	at com.aliyun.emr.rss.service.deploy.master.clustermeta.ha.HARaftServer.start(HARaftServer.java:280)
> 	at com.aliyun.emr.rss.service.deploy.master.clustermeta.ha.MetaHandler.setUpMasterRatisServer(MetaHandler.java:56)
> 	at com.aliyun.emr.rss.service.deploy.master.Master.<init>(Master.scala:53)
> 	at com.aliyun.emr.rss.service.deploy.master.Master$.main(Master.scala:643)
> 	at com.aliyun.emr.rss.service.deploy.master.Master.main(Master.scala)
> Caused by: java.net.BindException: Address already in use
> 	at sun.nio.ch.Net.bind0(Native Method)
> 	at sun.nio.ch.Net.bind(Net.java:438)
> 	at sun.nio.ch.Net.bind(Net.java:430)
> 	at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:225)
> 	at org.apache.ratis.thirdparty.io.netty.channel.socket.nio.NioServerSocketChannel.doBind(NioServerSocketChannel.java:134)
> 	at org.apache.ratis.thirdparty.io.netty.channel.AbstractChannel$AbstractUnsafe.bind(AbstractChannel.java:562)
> 	at org.apache.ratis.thirdparty.io.netty.channel.DefaultChannelPipeline$HeadContext.bind(DefaultChannelPipeline.java:1334)
> 	at org.apache.ratis.thirdparty.io.netty.channel.AbstractChannelHandlerContext.invokeBind(AbstractChannelHandlerContext.java:506)
> 	at org.apache.ratis.thirdparty.io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:491)
> 	at org.apache.ratis.thirdparty.io.netty.handler.logging.LoggingHandler.bind(LoggingHandler.java:230)
> 	at org.apache.ratis.thirdparty.io.netty.channel.AbstractChannelHandlerContext.invokeBind(AbstractChannelHandlerContext.java:506)
> 	at org.apache.ratis.thirdparty.io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:491)
> 	at org.apache.ratis.thirdparty.io.netty.channel.DefaultChannelPipeline.bind(DefaultChannelPipeline.java:973)
> 	at org.apache.ratis.thirdparty.io.netty.channel.AbstractChannel.bind(AbstractChannel.java:260)
> 	at org.apache.ratis.thirdparty.io.netty.bootstrap.AbstractBootstrap$2.run(AbstractBootstrap.java:356)
> 	at org.apache.ratis.thirdparty.io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
> 	at org.apache.ratis.thirdparty.io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
> 	at org.apache.ratis.thirdparty.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503)
> 	at org.apache.ratis.thirdparty.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
> 	at org.apache.ratis.thirdparty.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
> 	at org.apache.ratis.thirdparty.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
> 	at java.lang.Thread.run(Thread.java:748)
> 
> Process finished with exit code 255

**Comment by nafiyAix** [comment]
@FMX Thanks, understand, my thinking was a bit redundant before. According to the exception stack, I mistakenly thought that Ratis had caught all exceptions by below methods during election process then just print them into log. :dizzy_face:
```
     catch(Exception e) {
      final LifeCycle.State state = lifeCycle.getCurrentState();
      if (state.isClosingOrClosed()) {
        LOG.info("{}: {} is safely ignored since this is already {}",
            this, JavaUtils.getClassSimpleName(e.getClass()), state, e);
      } else {
        if (!server.getInfo().isAlive()) {
          LOG.info("{}: {} is safely ignored since the server is not alive: {}",
              this, JavaUtils.getClassSimpleName(e.getClass()), server, e);
        } else {
          LOG.error("{}: Failed, state={}", this, state, e);
        }
        shutdown();
      }
    } 
```

**Comment by nafiyAix** [comment]
ping @FMX @waitinfuture 
To keep the method reusability, I move the logic of System.exit() to the code block of the Master.
Below is the test results from my side.
when HA port is occupied and run start-master script
>starting com.aliyun.emr.rss.service.deploy.master.Master, logging to /var/log/rss/rss-root-com.aliyun.emr.rss.service.deploy.master.Master-1-dev01-hadoop.data-infra.shopee.io.out
failed to launch: nice -n 0 /usr/share/rss/bin/rss-class com.aliyun.emr.rss.service.deploy.master.Master
full log in /var/log/rss/rss-root-com.aliyun.emr.rss.service.deploy.master.Master-1-dev01-hadoop.data-infra.shopee.io.out

last output inside rss.log
>ERROR [main] Master: HA port 9872 of Ratis Server is occupied, please refer to configuration doc to modify the HA port in config file for each node
java.io.IOException: master1: Failed to start NettyRpcService
        at org.apache.ratis.netty.server.NettyRpcService.startImpl(NettyRpcService.java:141)
        at org.apache.ratis.util.LifeCycle.startAndTransition(LifeCycle.java:270)
        at org.apache.ratis.server.RaftServerRpcWithProxy.start(RaftServerRpcWithProxy.java:72)
        at org.apache.ratis.server.impl.RaftServerProxy.lambda$start$5(RaftServerProxy.java:375)
        at org.apache.ratis.util.LifeCycle.startAndTransition(LifeCycle.java:270)
        at org.apache.ratis.server.impl.RaftServerProxy.start(RaftServerProxy.java:373)
        at com.aliyun.emr.rss.service.deploy.master.clustermeta.ha.HARaftServer.start(HARaftServer.java:280)
        at com.aliyun.emr.rss.service.deploy.master.clustermeta.ha.MetaHandler.setUpMasterRatisServer(MetaHandler.java:52)
        at com.aliyun.emr.rss.service.deploy.master.Master.liftedTree1$1(Master.scala:58)
        at com.aliyun.emr.rss.service.deploy.master.Master.<init>(Master.scala:57)
        at com.aliyun.emr.rss.service.deploy.master.Master$.main(Master.scala:576)
        at com.aliyun.emr.rss.service.deploy.master.Master.main(Master.scala)
Caused by: java.net.BindException: Address already in use
        at sun.nio.ch.Net.bind0(Native Method)
        at sun.nio.ch.Net.bind(Net.java:433)
        at sun.nio.ch.Net.bind(Net.java:425)
        at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:223)
        at org.apache.ratis.thirdparty.io.netty.channel.socket.nio.NioServerSocketChannel.doBind(NioServerSocketChannel.java:134)
        at org.apache.ratis.thirdparty.io.netty.channel.AbstractChannel$AbstractUnsafe.bind(AbstractChannel.java:562)
        at org.apache.ratis.thirdparty.io.netty.channel.DefaultChannelPipeline$HeadContext.bind(DefaultChannelPipeline.java:1334)
        at org.apache.ratis.thirdparty.io.netty.channel.AbstractChannelHandlerContext.invokeBind(AbstractChannelHandlerContext.java:506)
        at org.apache.ratis.thirdparty.io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:491)
        at org.apache.ratis.thirdparty.io.netty.handler.logging.LoggingHandler.bind(LoggingHandler.java:221)
        at org.apache.ratis.thirdparty.io.netty.channel.AbstractChannelHandlerContext.invokeBind(AbstractChannelHandlerContext.java:506)
        at org.apache.ratis.thirdparty.io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:491)
        at org.apache.ratis.thirdparty.io.netty.channel.DefaultChannelPipeline.bind(DefaultChannelPipeline.java:973)
        at org.apache.ratis.thirdparty.io.netty.channel.AbstractChannel.bind(AbstractChannel.java:260)
        at org.apache.ratis.thirdparty.io.netty.bootstrap.AbstractBootstrap$2.run(AbstractBootstrap.java:356)
        at org.apache.ratis.thirdparty.io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
        at org.apache.ratis.thirdparty.io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
        at org.apache.ratis.thirdparty.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
        at org.apache.ratis.thirdparty.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at org.apache.ratis.thirdparty.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at org.apache.ratis.thirdparty.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)

**Review by FMX — APPROVED** [review]
LGTM.

**Pull request merged**